azure/services/inboundnatrules/inboundnatrules.go

```diff
@@ -19,136 +19,99 @@ package inboundnatrules
-		log.V(2).Info("creating inbound NAT rule", "NAT rule", inboundNatSpec.Name)
-
-		lb, err := s.loadBalancersClient.Get(ctx, s.Scope.ResourceGroup(), inboundNatSpec.LoadBalancerName)
-		if err != nil {
-			return errors.Wrapf(err, "failed to get Load Balancer %s", inboundNatSpec.LoadBalancerName)
-		}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-		if lb.LoadBalancerPropertiesFormat == nil || lb.FrontendIPConfigurations == nil || lb.InboundNatRules == nil {
-			return errors.Errorf("Could not get existing inbound NAT rules from load balancer %s properties", to.String(lb.Name))
-		}
-
-		ports := make(map[int32]struct{})
-		if s.natRuleExists(ports)(ctx, *lb.InboundNatRules, inboundNatSpec.Name) {
-			// Inbound NAT Rule already exists, nothing to do here.
-			continue
-		}
-
-		sshFrontendPort, err := s.getAvailablePort(ctx, ports)
-		if err != nil {
-			return errors.Wrapf(err, "failed to find available SSH Frontend port for NAT Rule %s in load balancer %s", inboundNatSpec.Name, to.String(lb.Name))
-		}
+	existingRules, err := s.client.List(ctx, s.Scope.ResourceGroup(), s.Scope.APIServerLBName())
+	if err != nil {
+		result := errors.Wrapf(err, "failed to get existing NAT rules")
+		s.Scope.UpdatePutStatus(infrav1.InboundNATRulesReadyCondition, serviceName, result)
+		return result
+	}
 
-		rule := network.InboundNatRule{
-			Name: to.StringPtr(inboundNatSpec.Name),
-			InboundNatRulePropertiesFormat: &network.InboundNatRulePropertiesFormat{
-				BackendPort:          to.Int32Ptr(22),
-				EnableFloatingIP:     to.BoolPtr(false),
-				IdleTimeoutInMinutes: to.Int32Ptr(4),
-				FrontendIPConfiguration: &network.SubResource{
-					ID: (*lb.FrontendIPConfigurations)[0].ID,
-				},
-				Protocol:     network.TransportProtocolTCP,
-				FrontendPort: &sshFrontendPort,
-			},
-		}
-		log.V(3).Info("Creating rule %s using port %d", "NAT rule", inboundNatSpec.Name, "port", sshFrontendPort)
+	existingRuleSpecs := make([]converters.ExistingInboundNatSpec, len(existingRules))
+	for i, rule := range existingRules {
+		existingRuleSpecs[i] = converters.InboundNatRuleToExistingInboundNatSpec(rule)
+	}
 
-		err = s.client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), to.String(lb.Name), inboundNatSpec.Name, rule)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create inbound NAT rule %s", inboundNatSpec.Name)
+	// We go through the list of InboundNatSpecs to reconcile each one, independently of the result of the previous one.
+	// If multiple errors occur, we return the most pressing one.
+	//  Order of precedence (highest -> lowest) is: error that is not an operationNotDoneError (i.e. error creating) -> operationNotDoneError (i.e. creating in progress) -> no error (i.e. created)
+	var result error
+	for _, natRule := range s.Scope.InboundNatSpecs(existingRuleSpecs) {
+		if _, err := s.CreateResource(ctx, natRule, serviceName); err != nil {
```

Comment:
thoughts for later: in theory if this was used to create multiple rules we'd have to update existingRules to include the previous ones that were just created since they are now taking ports that used to be available. This doesn't matter much in practice currently since we are only creating one rule per machine but could lead to errors down the road. Consider adding a TODO or a comment to that effect, or modifying the logic to include previously created rules in existing rules.